diem-move/parallel-executor/src/scheduler.rs

```diff
@@ -1,113 +1,470 @@
+// to 'ReadyToExecute(incarnation + 1)', allowing the scheduler to create an execution
+// task for the next incarnation of the transaction.
+//
+// Status transition diagram when with no dependencies:
+// ReadyToExecute(i) => --exactly once-- => Executing(i) => --transaction executed i-th time-- =>
+// => Executed(i) => --validations happen-- =>
+// 1. validation failures: --exactly once-- => Aborting(i) => => ReadyToExecute(i+1)
+// 2. no validation failures: status remains Executed(i).
+//
+// Status transition diagram when incarnation i observes a (read) dependency on
+// (the next incarnation of) transaction j:
+// ReadyToExecute(i) => --exactly once-- => Executing(i) => --suspended-- =>
+// => --txn j's next incarnation finishes execution-- => ReadyToExecute(i+1)
+#[derive(PartialEq)]
+enum TransactionStatus {
+    ReadyToExecute(Incarnation),
+    Executing(Incarnation),
+    Executed(Incarnation),
+    Aborting(Incarnation),
 }
 
 pub struct Scheduler {
-    // Shared index (version) of the next txn to be executed from the original transaction sequence.
-    execution_marker: AtomicUsize,
+    // A shared index that tracks the minimum of all transaction indices that require execution.
+    // The threads increment the index and attempt to create an execution task for the corresponding
+    // transaction, if the status of the txn is 'ReadyToExecute'. This implements a counting-based
+    // concurrent priority-queue. It is reduced as necessary when transactions become ready to be
+    // executed, in particular, when execution finishes and dependencies are resolved.
+    execution_idx: AtomicUsize,
+    // A shared index that tracks the minimum of all transaction indices that require validation.
+    // The threads increment the index and attempt to create a validation task for the corresponding
+    // transaction, if the status of the txn is 'Executed'. This implements a counting-based
+    // concurrent priority-queue. It is reduced as necessary when transactions require validation,
+    // in particular, after aborts and executions that write outside of the write set of the
+    // same transaction's previous incarnation.
+    validation_idx: AtomicUsize,
+    // The the number of times execution_idx and validation_idx are decreased.
+    decrease_cnt: AtomicUsize,
+
+    // Number of tasks used to track when transactions can be committed, incremented / decremented
+    // as new validation or execution tasks are created and completed.
+    num_active_tasks: AtomicUsize,
+    // Shared marker that is set when a thread detects that all txns can be committed.
+    done_marker: AtomicBool,
+
     // Shared number of txns to execute: updated before executing a block or when an error or
-    // reconfiguration leads to early stopping (at that transaction version).
-    stop_at_version: AtomicUsize,
+    // reconfiguration leads to early stopping (at that transaction idx).
+    stop_idx: AtomicUsize,
+    block_size: usize,
```

Comment:
I feel this duplicates the stop_idx

Reply:
got rid of all the uses of block size and removed it